cartographer/io/probability_grid_points_processor.cc

```diff
@@ -0,0 +1,147 @@
+#include "cartographer/io/probability_grid_points_processor.h"
+
+#include <cmath>
+#include <string>
+
+#include "Eigen/Core"
+#include "cairo/cairo.h"
+#include "cartographer/common/lua_parameter_dictionary.h"
+#include "cartographer/common/make_unique.h"
+#include "cartographer/common/math.h"
+#include "cartographer/io/image_utils.h"
+
+namespace cartographer {
+namespace io {
+namespace {
+
+struct PixelData {
+  uint8_t r = 0;
+  uint8_t g = 0;
+  uint8_t b = 0;
+};
+
+using PixelDataMatrix =
+  Eigen::Matrix<PixelData, Eigen::Dynamic, Eigen::Dynamic>;
+
+cairo_status_t CairoWriteCallback(void* const closure,
+                                  const unsigned char* data,
+                                  const unsigned int length) {
+  if (static_cast<FileWriter*>(closure)->Write(
+        reinterpret_cast<const char*>(data), length)) {
+    return CAIRO_STATUS_SUCCESS;
+  }
+  return CAIRO_STATUS_WRITE_ERROR;
+}
+
+void WriteImage(const PixelDataMatrix& mat, FileWriter* const file_writer) {
+  Image image = CreateImage(mat.cols(), mat.rows());
+
+  for (int y = 0; y < mat.rows(); ++y) {
+    for (int x = 0; x < mat.cols(); ++x) {
+      const PixelData& pixel = mat(y, x);
+      const int r = pixel.r;
+      const int g = pixel.g;
+      const int b = pixel.b;
+      image.pixels[y * image.stride / 4 + x] = (255 << 24) | (r << 16) | (g << 8) | b;
+    }
+  }
+
+  WritePng(&image, file_writer);
+  CHECK(file_writer->Close());
+}
+
+mapping_2d::ProbabilityGrid CreateProbabilityGrid(double resolution) {
+  constexpr int kInitialProbabilityGridSize = 100;
```

Comment:
What a terrible interface ProbabilityGrid provides for construction. I am sorry .... :(.